tests/common.rs

```diff
@@ -0,0 +1,273 @@
+///
+/// 1. CounterSubscriber:
+///     - a dummy subscriber that increments a counter on event
+///     - only uses one event
+///     - it has to be explicitly mutated; for this reason it implements `MutEventSubscriber`.
+///
+/// 2. CounterSubscriberWithData:
+///     - a dummy subscriber that increments a counter on events
+///     - this subscriber takes care of multiple events and makes use of `Events::with_data` so
+///       that in the `process` function it identifies the trigger of an event using the data
+///       instead of the file descriptor
+///     - it has to be explicitly mutated; for this reason it implements `MutEventSubscriber`.
+///
+/// 3. CounterInnerMutSubscriber:
+///     - a dummy subscriber that increments a counter on events
+///     - the subscriber makes use of inner mutability; multi-threaded applications might want to
+///       use inner mutability instead of having something heavy weight (i.e. Arc<Mutex>).
+///     - this subscriber implement `EventSubscriber`.
+use event_manager::{EventOps, EventSubscriber, Events, MutEventSubscriber};
+use vmm_sys_util::{epoll::EventSet, eventfd::EventFd};
+
+use std::fmt::{Display, Formatter, Result};
+use std::os::unix::io::AsRawFd;
+use std::sync::atomic::AtomicU64;
+use std::sync::atomic::Ordering;
+
+/// A `Counter` is a helper structure for creating subscribers that increment a value
+/// each time an event is triggered.
+/// The `Counter` allows users to assert and de-assert an event, and to query the counter value.
+pub struct Counter {
+    event_fd: EventFd,
+    counter: u64,
+}
+
+impl Display for Counter {
+    fn fmt(&self, f: &mut Formatter<'_>) -> Result {
+        write!(
+            f,
+            "(event_fd = {}, counter = {})",
+            self.event_fd.as_raw_fd(),
+            self.counter
+        )
+    }
+}
+
+impl Counter {
+    pub fn new() -> Self {
+        Self {
+            event_fd: EventFd::new(0).unwrap(),
+            counter: 0,
+        }
+    }
+
+    pub fn trigger_event(&mut self) {
+        let _ = self.event_fd.write(1);
+    }
+
+    pub fn clear_event(&self) {
+        let _ = self.event_fd.read();
+    }
+
+    pub fn counter(&self) -> u64 {
+        self.counter
+    }
+}
+
+impl Default for Counter {
+    fn default() -> Self {
+        Self::new()
+    }
+}
+
+// A dummy subscriber that increments a counter whenever it processes
+// a new request.
+pub struct CounterSubscriber(pub Counter);
+
+impl std::ops::Deref for CounterSubscriber {
+    type Target = Counter;
+
+    fn deref(&self) -> &Self::Target {
+        &self.0
+    }
+}
+
+impl std::ops::DerefMut for CounterSubscriber {
+    fn deref_mut(&mut self) -> &mut Counter {
+        &mut self.0
+    }
+}
+
+impl CounterSubscriber {
+    pub fn new() -> Self {
+        CounterSubscriber(Counter::new())
+    }
+}
+
+impl MutEventSubscriber for CounterSubscriber {
+    fn process(&mut self, events: Events, event_ops: &mut EventOps) {
+        match events.event_set() {
```

Comment:
In general, I think it's possible to receive multiple events at the same time (don't remember which combinations are valid and which aren't :( ). If that's the case, I think we should check for inclusion rather than equality when looking at the event_set maybe. Just raising this so we check what the best approach is. I guess this also applies to the other similar code snippets.

Reply:
Hmm, I'll double check that.